Blog on fine-tuning a model with KAITO (#4484)
Azure/AKS: blog/_posts/2024-08-23-KAITO-finetuning.md

---
title: "Fine tune language models with KAITO on AKS"
description: "Make your AI model even smarter, in just a few steps"
date: 2024-08-23
author: Sachi Desai

> categories: AI

Reviewer: ```suggestion
categories:
- ai
- add-ons
tags:
- kaito
```

---

## Introduction

You may have heard of the [Kubernetes AI Toolchain Operator](https://github.com/Azure/kaito/tree/main) (KAITO) announced at Ignite 2023 and KubeCon Europe this year. The open source project has gained popularity in recent months by introducing a streamlined approach to AI model deployment and flexible infrastructure provisioning on Kubernetes.

With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0), KAITO has expanded the supported model library to include the Phi-3 model, but the biggest (and most exciting) addition is the ability to fine-tune open-source models. Why should you be excited about fine-tuning? Well, it’s because fine-tuning is one way of giving your foundation model additional training using a specific dataset to enhance accuracy, which ultimately improves the interaction with end-users.

## Fine-tuning a foundation model is sometimes necessary

> Let’s experiment with [Phi-3-mini-128K](https://huggingface.co/microsoft/Phi-3-medium-128k-instruct), a robust and high-performing model that’s a bit smaller in size than your average LLM. We’ll use the following [chatbot UI tool](https://streamlit.io/), connected to a Phi-3 inference service deployed by KAITO, and ask a basic question about AKS:

Reviewer: ```suggestion
Let’s experiment with [Phi-3-Mini-128K-Instruct](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct), a robust and high-performing model that’s a bit smaller in size than your average LLM. We’ll use the following [Streamlit](https://streamlit.io/) chat UI app, connected to a Phi-3 inference service deployed by KAITO, and ask a basic question about AKS:
```

![Screenshot of untuned model response](/blog/assets/images/kaito_untuned_phi3_response.png)

> *Screenshot of un-tuned model response*

Reviewer: ```suggestion
_Screenshot of un-tuned model response_
```

Uh oh! That is **NOT** the answer we were expecting – we’d expect the model to know more about Kubernetes and cloud technologies. This foundation model was trained on a mix of synthetic data and public websites, so it doesn’t recognize AKS as Azure Kubernetes Service, and got creative in answering the question to the best of its ability. This is a well-known problem in the AI community and various fine-tuning techniques have been invented to enhance model accuracy using domain-specific knowledge. In our specific use case, we need the Phi-3 model to have more knowledge of Kubernetes and cloud platforms.

## KAITO can help with model fine-tuning

KAITO supports Low-Rank Adaption (LoRA) and Quantized LoRA (QLoRA), the [most common](https://arxiv.org/pdf/2402.16141) parameter-efficient fine-tuning techniques, which significantly lower the GPU requirement for tuning a language model. With LoRA, fine-tuning an LLM in a single GPU is possible! 

Let’s start the journey of fine-tuning the Phi-3 model using KAITO.

> First, we need an input dataset, so we’ll use a [public dataset](https://huggingface.co/datasets/sidddddddddddd/kubernetes) that is optimized for knowledge about Kubernetes and cloud platforms. (Note: this is the original data set which we copied and reformatted using [these instructions](https://github.com/Azure/kaito/tree/main/docs/tuning#input-dataset-format) from the HuggingFace trainer library.) 

Reviewer: ```suggestion
First, we need an input dataset, so we’ll use a [public dataset](https://huggingface.co/datasets/sidddddddddddd/kubernetes) that is optimized for knowledge about Kubernetes and cloud platforms.

> This is the original dataset which we selected and reformatted using [these instructions](https://github.com/Azure/kaito/tree/main/docs/tuning#input-dataset-format) from the HuggingFace trainer library.
```

> To use the data set in KAITO, we can specify it in the new `tuning` property in the workspace API. 

Reviewer: ```suggestion
To use the data set in KAITO, we can specify it in the new **tuning** property in the workspace API. 
```

> Here’s an example of what the fine-tuning workspace looks like when using a public dataset from HuggingFace:

Reviewer: ```suggestion
Here’s an example of what the fine-tuning workspace looks like when using a public dataset from HuggingFace:

```

**Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).

**Output**: where the adapter is stored, as a container image or any other storage type supported by Kubernetes.

> The choice of GPU SKU is critical since model fine-tuning normally requires more GPU memory compared to model inference. To avoid GPU Out-Of-Memory, NVIDIA A100 or higher tier GPUs are recommended. 

You can check out the [KAITO tuning API](https://github.com/Azure/kaito/tree/main/docs/tuning) for more details and troubleshooting tips!

After deploying the tuning workspace, KAITO will create a Job workload in the same namespace as the workspace and run to completion. A common perception is that model training job would take a long time – well, yes it can be multiple hours even for fine-tuning. You can track the tuning progress in the Job pod log, reported by the number of steps completed out of the total. 

> If you are curious to know how the total steps are calculated, here is the formula: 
> `total steps = number of epochs * (number of samples in dataset / batch size)`, where the number of epochs and batch size are configurable parameters. 

> KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and require better GPUs in general. 

Reviewer: ```suggestion
KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and may require different GPU VM SKUs in general. 
```

> When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be pulled into a new inferencing service!

Reviewer: ```suggestion
When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be added to new inferencing workspaces!
```

After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

![Three hours later](/blog/assets/images/three_hours_later_image.png)

> ## Now that tuning is done, how can we quickly see the result?

Reviewer: ```suggestion
## Better results with fine-tuned workspaces
```

To use our new fine-tuned adapter, the inference workspace will now accept one or more adapters to customize model behavior.

Here is an example of what a fine-tuned inference workspace looks like:

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-phi-3-mini-adapter
resource:
  instanceType: Standard_NC6s_v3
  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
```

> As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 

Reviewer: ```suggestion
As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 

```

> Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:
> 

Reviewer: Removing extra line break

```suggestion
Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:
```

![Screenshot of tuned model response](/blog/assets/images/kaito_tuned_phi3_response.png)

> *Screenshot of tuned model response*

Reviewer: ```suggestion
_Screenshot of tuned model response_
```

That definitely looks more accurate – the fine-tuned model quickly discerned that we’re looking for an answer related to Kubernetes and provided a much better response. However, there’s always room to improve the model’s accuracy for domain specific questions. We can improve upon or find new input datasets to create new tuning adapters. This iterative tuning process can be done by repeating the steps described above.

## Other options to make models smarter

> Retrieval-Augmented Generation (RAG) is another common technique used to improve the inference accuracy of foundation models. Compared to LoRA fine-tuning, RAG eliminates the need of training Jobs to generate adapters. It has a more complicated workflow with additional components like a vector database, indexing and query servers, etc. The good news is that RAG support with a simple user experience is in KAITO’s roadmap. 🚀

Reviewer: ```suggestion
Retrieval-Augmented Generation (RAG) is another common technique used to improve the inference accuracy of foundation models. Compared to LoRA fine-tuning, RAG eliminates the need of training jobs to generate adapters. It has a more complicated workflow with additional components like a vector database, indexing and query servers, etc. The good news is that RAG support with a simple user experience is in KAITO’s roadmap. 🚀
```

> If you are curious to know how the total steps are calculated, here is the formula: 
> `total steps = number of epochs * (number of samples in dataset / batch size)`, where the number of epochs and batch size are configurable parameters. 

KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and may require higher performing GPUs in general. 

When the fine-tuning job is complete, the result, which is often referred to as an **adapter**, will be packaged as a container image and stored in the specified output location. The adapter image is lightweight, portable, conveniently version controlled, and can be added to new inferencing workspaces!

After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

> ![Three hours later](/blog/assets/images/three_hours_later_image.png)

Reviewer: ```suggestion
![Three hours later](/blog/assets/images/three_hours_later_image.png)

> This particular fine-tuning job took a little over 3 hours, but job completion duration will vary depending on the foundation model and the size of the input dataset.

```

**Input**: for flexibility, the tuning input can be a URL (for a public dataset), or images (for a private dataset).

**Output**: where the adapter is stored, as a container image or any other storage type supported by Kubernetes.

> The choice of GPU SKU is critical since model fine-tuning normally requires more GPU memory compared to model inference. To avoid GPU Out-Of-Memory, NVIDIA A100 or higher tier GPUs are recommended. 

You can check out the [KAITO tuning API](https://github.com/Azure/kaito/tree/main/docs/tuning) for more details and troubleshooting tips!

After deploying the tuning workspace, KAITO will create a Job workload in the same namespace as the workspace and run to completion. A common perception is that model training job would take a long time – well, yes it can be multiple hours even for fine-tuning. You can track the tuning progress in the Job pod log, reported by the number of steps completed out of the total. 

> If you are curious to know how the total steps are calculated, here is the formula: 
> `total steps = number of epochs * (number of samples in dataset / batch size)`, where the number of epochs and batch size are configurable parameters. 

> KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and require better GPUs in general. 

Author: ```suggestion
KAITO allows users to apply a custom ConfigMap to overwrite most of the tuning parameters used by the tuning job. We should be careful about changing those parameters though. For example, increasing the batch size, i.e., leveraging higher data parallelism, will reduce the tuning time but use more GPU memory, and may require higher performing GPUs in general. 
```

After submitting your tuning workspace manifest, go get yourself a drink or a snack. 😉

![Three hours later](/blog/assets/images/three_hours_later_image.png)

## Better results with fine-tuned workspaces

To use our new fine-tuned adapter, the inference workspace will now accept one or more adapters to customize model behavior.

Here is an example of what a fine-tuned inference workspace looks like:

```bash
apiVersion: kaito.sh/v1alpha1
kind: Workspace
metadata:
  name: workspace-phi-3-mini-adapter
resource:
  instanceType: Standard_NC6s_v3
  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
```

As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 

Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:

> 

Author: ```suggestion
```

With the [v0.3.0 release](https://github.com/Azure/kaito/releases/tag/v0.3.0), KAITO has expanded the supported model library to include the Phi-3 model, but the biggest (and most exciting) addition is the ability to fine-tune open-source models. Why should you be excited about fine-tuning? Well, it’s because fine-tuning is one way of giving your foundation model additional training using a specific dataset to enhance accuracy, which ultimately improves the interaction with end-users.

## Fine-tuning a foundation model is sometimes necessary

Let’s experiment with [Phi-3-Mini-128K-Instruct](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct), a robust and high-performing model that’s a bit smaller in size than your average LLM. We’ll use a custom [Streamlit](https://streamlit.io/) chat UI app, connected to a Phi-3 inference service deployed by KAITO, and ask a basic question about AKS:

> ![Screenshot of untuned model response](/blog/assets/images/kaito_untuned_phi3_response.png)

Author: ```suggestion
![Screenshot of untuned model response](/AKS/assets/images/kaito-blog/untuned-lm-with-kaito.png)
```

  labelSelector:
    matchLabels:
      apps: phi-3-adapter
inference:
  preset:
    name: phi-3-mini-128k-instruct
  adapters:
    - source:
        name: myadapter
        image: myregistry.azurecr.io/adapters/myadapter:0.0.1
        imagePullSecrets:
          - myregistrysecret
```

As new inference requests come in, they’ll flow through the model that merges the adapters and the response will be affected by the newly trained data. 

Great! After updating our inference endpoint to leverage the new adapter, let’s ask the same question and check the response from the fine-tuned phi-3-mini-128k-instruct model:

![Screenshot of tuned model response](/blog/assets/images/kaito_tuned_phi3_response.png)
_Screenshot of tuned model response_

That definitely looks more accurate – the fine-tuned model quickly discerned that we’re looking for an answer related to Kubernetes and provided a much better response. However, there’s always room to improve the model’s accuracy for domain specific questions. We can improve upon or find new input datasets to create new tuning adapters. This iterative tuning process can be done by repeating the steps described above.

> ## Other options to make models smarter

Author: ```suggestion
## "Stay tuned" to make models smarter
```
